### PR TITLE
Pod::Wordlist: use File::Share instead of File::ShareDir::ProjectDistDir

### PR DESCRIPTION
**Pull request by dolmen**
@kentfredric Is it what you want?

### CODE REVIEW

**Comment by dolmen** [comment]
Fixes #19.

**Comment by kentfredric** [comment]
F:SD:PDD  added a defensive mechanism to avoid this problem by requiring the share dir in question **ALSO** contained a subdir matching the project name, because the incidence of that was significantly lower risk. ( This is referenced in documentation as "Strict mode", which xeno was using )

Its better, but the entire mentality of "decide during runtime whether we're installed or not by looking at the filesystem" is just all fail. There is no sane way of doing it that I have seen yet, and nobody has been able to provide one.

The "recommended" choice that I would recommend is not doing any runtime magic, and instead, using Pure File::ShareDir in the module itself, and then being *explicit* in tests and runtime that you are wanting to source from the development tree, using a tool like Test::File::ShareDir to get `File::ShareDir` to see it.

Sadly, this is presently somewhat hamstrung because doing that relies on tweaking `@INC` by cloning stuff into a tempdir, as that is the only viable way to work around the defects in `File::ShareDir`'s API which I have been **begging** to get fixed for several years.

**Comment by kentfredric** [comment]
Actual example of a failure caused by the logic: https://github.com/kentnl/File-ShareDir-ProjectDistDir/issues/7

**Comment by karenetheridge** [comment]
> The "recommended" choice that I would recommend is not doing any runtime magic, and instead, using Pure File::ShareDir in the module itself, and then being explicit in tests and runtime that you are wanting to source from the development tree, using a tool like Test::File::ShareDir to get File::ShareDir to see it.

I agree. Keep the runtime simple, and do the special swapping only in tests. Test::File::ShareDir is a bit icky on the inside, but this is fixable, and it's better to have the weird code in tests than at runtime.

**Comment by kentfredric** [comment]
Also, bonus guesswork to ask what happens if anyone tries to use Pod::Wordlist without loading Pod::Spell... https://metacpan.org/source/INGY/File-Share-0.25/lib/File/Share.pm#L27

I expect carnage.

**Comment by kentfredric** [comment]
> Test::File::ShareDir is a bit icky on the inside

The least icky ways involve using either the specific subclasses ( which are more optimised for individual usecases ), or using the OO guts underneath.

I hope to make it much nicer when there's a real API we can hook, but atm its just "*sigh*, create a tempdir, fill it with our tears, shuffle our feet as we stuff it in `@INC`, pretend its ok, eat chocolate and feel better"

**Comment by xenoterracide** [comment]
There was a goal to not have Pod::Wordlist expose a globally scoped hash (requires fixing external modules) and have the wordlist file injected into the wordlist, so that we didn't have to have everyone making Pod::Wordlist::MyWordlist but could instead just provide the file, of course it still then needs a default that ships and installs.

**Comment by xenoterracide** [comment]
actually... injecting the file would solve the heuristic problem I think... but you'd have to not use the sharedir api, which is broken and why you have to modify @Inc. Instead add a feature to Path::Tiny (or whatever) where you return the Path::Tiny::File object, have a special method ->sharedir('file') that will use the actual sharedir, have Pod::Wordlist take a Path::Tiny::File as a parameter, have Pod::Spell provide it. Have Pod::Spell take a parameter of either a wordlist path string, or a Path::Tiny::File. In tests then simply inject the test location of Path::Tiny::File, no more hacking @Inc and no more runtime heuristics. Dependency Inversion++

**Comment by kentfredric** [comment]
Can you rephrase that last comment? I'm having an incredibly hard time understanding what you're suggesting. ( Also, Path::Tiny::File is not a thing, all paths are just Path::Tiny instances, and there's no file/dir separation )

**Comment by xenoterracide** [comment]
~~~
Pod::Spell->new(
    wordlist => Pod::Wordlist->new(
          path => Path::Tiny::File->sharedir('Pod-Spell','wordlist') # Path::Tiny::File
     )
)
~~~
in tests
~~~
Pod::Spell->new(
    wordlist => Pod::Wordlist->new(
          path => Path::Tiny::File->file('share/wordlist') # Path::Tiny::File
     )
)
~~~
by default use the first internally, but allow the wordlist with the path provided.

**Comment by kentfredric** [comment]
I don't see why Path::Tiny is even needed there.

You could just do:
```perl
Pod::Spell->new(
    wordlist => Pod::Wordlist->new(
          path => dist_file('Pod-Spell', 'wordlist')
     )
)
```

That is, assume the API takes a string or something that stringifies, and put the burden on the consumer to get it right. ( And they can thus use whatever glue they want ).

Making Path::Tiny have sharedir behaviour is just putting logic into the wrong place.

You could make it sweeter and have: 
```perl
Pod::Spell->new(    wordlist => ':default',)
```

Which would be some convenience glue for people who never needed `./share/` stuff.

Additionally, the following may be a tempting idea:


```perl
Pod::Spell->new(
    wordlist => Pod::Wordlist->new(
          paths => [
                  dist_file('Pod-Spell', 'wordlist'),
                  dist_file('My-Dist', 'wordlist'),
         ]
     )
)
```

Which would be written as:

```perl
Pod::Spell->new(
    wordlist => Pod::Wordlist->new(
          paths => [
                  dist_file('Pod-Spell', 'wordlist'),
                  path("./share/wordlist")
         ]
     )
)
```

In the dists tests themselves.

This means test authors also can choose their mechanism of accessing said files, either directly, or by using Test::File::ShareDir if they so figured it necessary.

**Comment by xenoterracide** [comment]
maybe you don't (it's first thing in the morning and I don't do perl anymore)? I guess whether a stringy path to the file is good enough... I'll leave that up to you, but my idea is basically don't rely internally on sharedir or the test sharedir and the @Inc modifications, allow the client of Pod::Spell to do that.

**Comment by kentfredric** [comment]
The only really big disadvantage of this approach is it requires every consumer to know the name of your files you ship in your share dir. Which I find a bit horrible.

**Comment by xenoterracide** [comment]
that's why you provide a default to the parameter, so it doesn't have to be supplied.

**Comment by kentfredric** [comment]
I feel like its the sort of thing that needs a registration node as a module, say for instance, 
 
(pseudocode)
```perl
package Pod::Wordlist::Default;

sub filename { 'wordlist' }

...

# Some other module
sub _wordlist_file_path {
    my ( $name ) = @_;
    my $module = "Pod::Wordlist::" . $name; #
    eval q[ require $module ];
    # Allow dictionaries to use whatever technique they want as long as they 
    # can sort it out themselves
    return $module->path if $module->can('path');
    die "Can't work out how to load $module" if not $module->can('filename');
    # allow some sugar that if they declare a sub named "distname" and a sub called"filename"
    # then we can divine it from File::ShareDir as a dist-file
    return dist_file( $module->distname, $module-> filename ) if $module->can('distname');
    my $share_module = $module;
    $share_module = $module->modulename if $module->can('modulename'); # YAGNI really.
    # Otherwise assume that the module name itself is the anchor for the shared file.
    return module_file( $share_module, $module->filename );
}

...
# In code
Pod::Wordlist->new(
    dictionaries => [qw( Default  MyDictionary )],
);
# In the tests for dist containing MyDictionary
Pod::Wordlist->new(
    dictionaries => [qw( Default )],
    files => ['./share/wordlist'],
);
```

And you could sugar syntax the whole process via:

```perl
Pod::Spell->new(
    wordlist => [qw( Default  MyDictionary )]   # shorthand to create a wordlist object
);
```

**Comment by xenoterracide** [comment]
I mean, in all honesty ::Wordlist itself is kind of that module. with your sugar it seems like you'd end right back up at not providing the path which results in the same problem, the path not being external to the object, making things like testing harder, or providing your own path for whatever reason.

 the only reason I can think of to have different ::Wordlist objects is to have different types of file sources e.g. Wordlist::Aspell, or Wordlist::CSV

 Either way I think this is going down the right path of not having to hack the @Inc in order to write tests (or some other magic that changes the internal method call of my class), and putting more power in the hands of the consumer

**Comment by kentfredric** [comment]
> Either way I think this is going down the right path of not having to hack the @Inc in order to write tests (or some other magic that changes the internal method call of my class), and putting more power in the hands of the consumer

That's entirely the point.

You wouldn't.

The infrastructure I laid out above means people who are using the word list outside the dist don't need to know where its stored.

And people who are bundling wordlists in their own dists can refer to them by direct path ( without need for a registration node )

Whereas your proposal gives only the latter, and forces consumers to refer to them by direct path even when they're not operating in the context of a wordlist author.

The dual-pronged approach means you can pick whichever mechanism you want to refer to the file paths. You can specify a dictionary using an `@INC` locatable module name, or you can specify a dictionary using a file path. The `@INC` locatable module name is purely ease-of-use for people who wish to consume said dictionaries, not author them.

Because referring to a dictionary by its *filename* is essentially a hidden dependency, a dependency CPAN dependency systems can't handle. So instead, by referring to them by module name, you open a door to being able to manage the dependency sanely when it is an external dependency, and the module in question has veto power over how it finds its path.

Hence: 
```perl
Pod::Wordlist->new(
    dictionaries => [qw( Default )],
    files => ['./share/wordlist'],
);
```

Which is just going to be sugar syntax for:

```perl
# In My-Dist's tests
Pod::Wordlist->new(
    files => [ 
        module_file('Pod::Wordlist::Default','wordlist'), 
       './share/wordlist'
    ],
);
```

Or:
```perl
# In My-Dist's tests
Pod::Wordlist->new(
    files => [ 
       distfile_file(Pod::Wordlist::Default->distname,'wordlist'), 
       './share/wordlist'
    ],
);
```

**Comment by dolmen** [comment]
Fixed instead by #26.
